File: CODE/perm2.f90
! Perm function 2
! -n <= x <= n
! fstar = 0
!--------------------
!     PERM2 FUNCTION
!--------------------
      subroutine funct(n,x,f)
      implicit none
      integer j,k,n
      real*8 f,x(n),s
      f=0.d0
      do k=1,n
       s=0.d0
       do j=1,n
        s=s+(j+10.d0)*(x(j)**k-(1.d0/j)**k)
       end do
       f=f+s**2
      end do
      return
      end
!-------------
!     GRADIENT
!-------------
      subroutine grad(n,x,g)
      implicit none
      integer i,j,k,n
      real*8 x(n),g(n),s1,s
      s1=0.d0
      do j=1,n
       s1=s1+(j+10.d0)*(x(j)-1.d0/j)
      end do
      do i=1,n
       g(i)=2.d0*s1*(i+10.d0)
       do k=2,n
        s=0.d0
        do j=1,n
         s=s+(j+10.d0)*(x(j)**k-(1.d0/j)**k)
        end do
        g(i)=g(i)+2.d0*s*k*(i+10.d0)*x(i)**(k-1)
       end do
      end do
      return
      end

File: Python/perm2.py
# Perm function 2
# -n <= x <= n
# fstar = 0
#--------------------
#     PERM2 FUNCTION
#--------------------
import numpy as np

def perm2_function(x):
    # PERM2 function
    n = len(x)
    f = 0.0

    for k in range(1, n + 1):
        s = 0.0
        for j in range(1, n + 1):
            s += (j + 10.0) * (x[j-1]**k - (1.0/j)**k)
        f += s**2

    return f

def perm2_gradient(x):
    # Gradient of the PERM2 function
    n = len(x)
    g = np.zeros(n)
    s1 = 0.0

    for j in range(1, n + 1):
        s1 += (j + 10.0) * (x[j-1] - 1.0/j)

    for i in range(1, n + 1):
        g[i-1] = 2.0 * s1 * (i + 10.0)

        for k in range(2, n + 1):
            s = 0.0
            for j in range(1, n + 1):
                s += (j + 10.0) * (x[j-1]**k - (1.0/j)**k)
            
            g[i-1] += 2.0 * s * k * (i + 10.0) * x[i-1]**(k - 1)

    return g

if __name__ == "__main__":
    # Example usage
    x_example = np.zeros(4)

    function_value = perm2_function(x_example)
    gradient_value = perm2_gradient(x_example)

    print("Function value:", function_value)
    print("Gradient value:", gradient_value)
